### ISSUE TITLE
Loading followers from other instances

### ISSUE THREAD

**Issue by MariaFeodora** [Dimillian/IceCubesApp#195]

**Comment by Dimillian** [comment]
Unfortunately that will not be possible but I'll add a message.

**Comment by fifafu** [comment]
Just checking: why do you think this won't be possible? The Mona client is using these API's which are publicly accessible on each instance:

https://docs.joinmastodon.org/methods/accounts/#lookup 
https://docs.joinmastodon.org/methods/accounts/#following
https://docs.joinmastodon.org/methods/accounts/#followers

**Comment by Dimillian** [comment]
There is a high chance it can be private and on top of that, it'll show you the followers of the user on your instance and their instance but what about the potential 10000 other instances?

**Comment by fifafu** [comment]
I have not encountered a single instance where it was private.
It doesn't only show you the followers on that instance, it shows you all followers it knows about. And because the user must have followed these accounts, the instance knows all of them.

It's  the same API that is used in the web interface.

**Comment by Dimillian** [comment]
I don't understand. It's already what the app is doing? If you tap on a user followers / following. You can see the list. And on the web they added a label to tell you they don't show users from other instances.

**Comment by Dimillian** [comment]
Maybe I've missed something.

**Comment by Dimillian** [comment]
Not sure we want to go in this direction. It feels quite heavy compared to the philosophie of the app. I do want to have local storage and caching but I'm not there yet. I've explored a few solutions this weekend for the timeline.

**Comment by fifafu** [comment]
Let's say I open the IceCubes app and I'm logged in with **my** instance (troet.cafe), and I want to see the people you (dimillian@mastodon.scoail) follow (or those who follow you).  Currently the app only shows a handful of accounts because IceCubs does **not** try to load them from your instance (which would be mastodon.social), but instead it tries to load them from the instance I'm logged in with (troet.cafe).

To solve this I think it should call these, even if I'm logged in with troet.cafe.
1.) https://mastodon.social/api/v1/accounts/lookup?acct=dimillian
2.) Use the returned it to call https://mastodon.social/api/v1/accounts/109523208135675669/followers

**Comment by Dimillian** [comment]
Totally make sense! Will add that. It's clearly an oversight that it's not loading people from the user current instance. Thanks for explaining.

**Comment by fifafu** [comment]
That would be awesome and it would definitely make Ice Cubes my favorite Mastodon app 👍 It's the only thing I have been missing (especially because I'm on a smaller instance)

**Comment by Dimillian** [comment]
Feel free. Don't really know what we can do about it

**Comment by fifafu** [comment]
Just in case somebody wants to work on this, a concrete example of how it would need to be done:

1.) Lookup the user id by querying the user's instance with his username: 
```https://mastodon.social/api/v1/accounts/lookup?acct=dimillian```

2.) Use the user id to call the accounts/:userid/followers endpoint
```
https://mastodon.social/api/v1/accounts/109523208135675669/followers?limit=80
```

3.) Use paging to load more:
```
https://mastodon.social/api/v1/accounts/109523208135675669/followers?max_id=51115796&limit=80
```

This API is not protected, so it can be called without any authentication.

**Comment by pshadov** [comment]
Unfortunately, after digging deeper into the problem, I found that it is not so straightforward and requires many more changes.

To make it shorter, I will call the instance I am logged in MI (my instance) and the instance of the user I am looking at TI (target instance)

I identified 3 problems (I assume there might be more down the road):
1. The relationships with the followed/following users. If I get the list of the users on TI, their IDs will be different there comparing to MI. For example, @Dimillian ID on `mastodon.social` is **109523208135675669**, but on `mas.to` - **109523215128421330**. So basically, I have a nice list of all the users from TI, but I don't know if I am already following them on MI, and I haven't found any API for the user or relationships IDs mapping; 
2. Let's imagine I press on some user from the followers list. Their ID is from TI. This means that without the mentioned IDs mapping, l cannot get all the information related to MI on their Profile screen;
3. I haven't gone too far into the codebase yet, but it looks to me that the main business logic of how Mastodon client passed in the app is coupled to SwiftUI environments. Making the logic more friendly to multiple clients (MI and TIs) can require a lot of architectural changes that can lead to more bugs.

I will dig deeper into the codebase and MastodonAPI, as I find them very interesting overall, but so far, I cannot promise that I will the issue soon. You can find the POC with disabled relationships but working followers/following lists on the branch on my fork [here](https://github.com/pshadov/IceCubesApp/tree/followers-according-to-the-server).

**Comment by fifafu** [comment]
Yes that's true. Other clients continue by resolving the user via /api/v2/search?resolve=true&q=dimillian%40mastodon.social&type=accounts&limit=1 (on the logged in instance)

However this only works for point 2. Point 1 is probably hard to solve.

**Comment by Dimillian** [comment]
I'm using the resolver for loading remote timelines and embedding remote statuses. But yes, it's an id by id case. No way to do it on lists of entities.
